osquery/tables/system/linux/model_specific_register.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <dirent.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <osquery/core.h>
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#define MSR_FILENAME_BUFFER_SIZE 32
+
+#define NO_MASK 0xFFFFFFFFFFFFFFFFULL
+
+// Defines taken from uapi/asm/msr-index.h from the linux kernel.
+#define MSR_PLATFORM_INFO 0x000000ce
+
+#define MSR_IA32_FEATURE_CONTROL 0x0000003a
+
+#define MSR_IA32_PERF_STATUS 0x00000198
+#define MSR_IA32_PERF_CTL 0x00000199
+#define INTEL_PERF_CTL_MASK 0xffff
+
+#define MSR_IA32_MISC_ENABLE 0x000001a0
+
+#define MSR_TURBO_RATIO_LIMIT 0x000001ad
+
+#define MSR_IA32_MISC_ENABLE_TURBO_DISABLE_BIT 38
+#define MSR_IA32_MISC_ENABLE_TURBO_DISABLE \
+  (1ULL << MSR_IA32_MISC_ENABLE_TURBO_DISABLE_BIT)
+
+// Run Time Average Power Limiting (RAPL).
+#define MSR_RAPL_POWER_UNIT 0x00000606
+#define MSR_PKG_ENERGY_STATUS 0x00000611
+#define MSR_PKG_POWER_LIMIT 0x00000610
+
+namespace osquery {
+namespace tables {
+
+// These are the entries to retrieve from the model specific register
+struct msr_record_t {
+  const char *name;
+  const off_t offset;
+  const uint64_t mask;
+  const int is_flag;
+};
+const static msr_record_t fields[] = {
+    {.name = "turbo_disabled",
+     .offset = MSR_IA32_MISC_ENABLE,
+     .mask = MSR_IA32_MISC_ENABLE_TURBO_DISABLE,
+     .is_flag = true},
+    {.name = "turbo_ratio_limit",
+     .offset = MSR_TURBO_RATIO_LIMIT,
+     .mask = NO_MASK,
+     .is_flag = false},
+    {.name = "platform_info",
+     .offset = MSR_PLATFORM_INFO,
+     .mask = NO_MASK,
+     .is_flag = false},
+    {.name = "perf_status",
+     .offset = MSR_IA32_PERF_STATUS,
+     .mask = NO_MASK,
+     .is_flag = false},
+    {.name = "perf_ctl",
+     .offset = MSR_IA32_PERF_CTL,
+     .mask = INTEL_PERF_CTL_MASK,
+     .is_flag = false},
+    {.name = "feature_control",
+     .offset = MSR_IA32_FEATURE_CONTROL,
+     .mask = NO_MASK,
+     .is_flag = false},
+    {.name = "rapl_power_limit",
+     .offset = MSR_PKG_POWER_LIMIT,
+     .mask = NO_MASK,
+     .is_flag = false},
+    {.name = "rapl_energy_status",
+     .offset = MSR_PKG_ENERGY_STATUS,
+     .mask = NO_MASK,
+     .is_flag = false},
+    {.name = "rapl_power_units",
+     .offset = MSR_RAPL_POWER_UNIT,
+     .mask = NO_MASK,
+     .is_flag = false}};
+
+void getModelSpecificRegisterData(QueryData &results, int cpu_number) {
+  char msr_filename[MSR_FILENAME_BUFFER_SIZE];
+  int length = snprintf(msr_filename, MSR_FILENAME_BUFFER_SIZE,
```

Comment:
You can use `std::to_string` and `std::string::operator+` instead of `snprintf` if you want more C++11 points.

```cpp
auto msr_filename = std::string("/dev/cpu/") + std::to_string(cpu_number) + "/msr";
```